ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java

```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
                             + ((trans != null) ? trans.length : 0)
                             + " xids in doubt");
                 }
-				if (jtaLogger.logger.isTraceEnabled()) {
-					for (Xid xid : trans) {
-						byte[] globalTransactionId = xid.getGlobalTransactionId();
-						byte[] branchQualifier = xid.getBranchQualifier();
-	
-						StringBuilder stringBuilder = new StringBuilder();
-						stringBuilder.append("< ");
-						stringBuilder.append(xid.getFormatId());
-						stringBuilder.append(", ");
-						stringBuilder.append(globalTransactionId.length);
-						stringBuilder.append(", ");
-						stringBuilder.append(branchQualifier.length);
-						stringBuilder.append(", ");
-						for (int i = 0; i < globalTransactionId.length; i++) {
-							stringBuilder.append(globalTransactionId[i]);
-						}
-						stringBuilder.append(", ");
-						for (int i = 0; i < branchQualifier.length; i++) {
-							stringBuilder.append(branchQualifier[i]);
-						}
-						stringBuilder.append(" >");
-	
-						jtaLogger.logger.debug("Recovered: "
-								+ stringBuilder.toString());
-					}
-				}
-			}
+
+                if (trans != null) {
+                    List<Xid> transList = new ArrayList<Xid>(Arrays.asList(trans));
+                    Iterator<Xid> it = transList.iterator();
+                    while (it.hasNext()) {
+                        Xid xid = it.next();
+                        if (xid != null) {
+                            byte[] globalTransactionId = xid.getGlobalTransactionId();
+                            if (globalTransactionId != null) {
```

Comment:
A topic for discussion - would make sense to check for `null` the format and branch as well? Is there some experience that e.g. Oracle could not be filling the other parts of Xid and leaving them `null`?

Reply:
I added check for branch == null.